lib/compiler.js

```diff
@@ -288,63 +289,63 @@ function getFileDependencies (compiler) {
 }
 
 /**
- * @type {WeakMap<WebpackCompilation, WeakMap<HtmlWebpackChildCompiler, boolean>>}}
+ * @type {WeakMap<object, Snapshot>}
  */
-const hasOutdatedCompilationDependenciesMap = new WeakMap();
+const snapshotMap = new WeakMap();
+const fileDepependenciesKeyRef = {};
 /**
- * Returns `true` if the file dependencies of the current childCompiler
- * for the given mainCompilation are outdated.
+ * Calls the callback with `true` if the file dependencies of the current childCompiler
+ * for the given mainCompilation are valid.
  *
- * Uses the `hasOutdatedCompilationDependenciesMap` cache if possible.
+ * Uses the `snapshotMap` cache if possible.
  *
  * @param {WebpackCompilation} mainCompilation
- * @returns {boolean}
+ * @param callback
  */
-function hasOutDatedTemplateCache (mainCompilation) {
+function hasOutDatedTemplateCache (mainCompilation, callback) {
   const childCompiler = getChildCompiler(mainCompilation.compiler);
-  /**
-   * @type {WeakMap<HtmlWebpackChildCompiler, boolean>|undefined}
-   */
-  let hasOutdatedChildCompilerDependenciesMap = hasOutdatedCompilationDependenciesMap.get(mainCompilation);
-  // Create map for childCompiler if none exist
-  if (!hasOutdatedChildCompilerDependenciesMap) {
-    hasOutdatedChildCompilerDependenciesMap = new WeakMap();
-    hasOutdatedCompilationDependenciesMap.set(mainCompilation, hasOutdatedChildCompilerDependenciesMap);
+
+  const keyStr = JSON.stringify([...childCompiler.fileDependencies].sort());
+  // get reference for key-strings to use it with WeakMap:
+  let key = fileDepependenciesKeyRef[keyStr];
+  if (!key) {
+    key = {};
+    fileDepependenciesKeyRef[keyStr] = key;
```

Comment:
This is a memory leak as it will grow on every compilation if output name contains a hash.